File: app/src/main/java/com/konektedi/vs/News/NewsAdapter.java
package com.konektedi.vs.News;


import android.content.Context;
import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.konektedi.vs.R;

import java.util.List;

public class NewsAdapter extends RecyclerView.Adapter<NewsAdapter.ViewHolder> {

    private Context mContext;
    private List<NewsModel> newsList;

    NewsAdapter(Context mContext, List<NewsModel> newsList) {
        this.mContext = mContext;
        this.newsList = newsList;
    }

    @NonNull
    @Override
    public NewsAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view;
        view = LayoutInflater.from(mContext).inflate(R.layout.z_post, parent, false);

        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(NewsAdapter.ViewHolder holder, final int position) {
        holder.titleView.setText(newsList.get(position).getTitle());
        holder.timeView.setText(newsList.get(position).getTime());
        holder.nameView.setText(newsList.get(position).getName());

        holder.titleView.setOnClickListener(view -> {
            Intent intent = new Intent(mContext, NewsView.class);
            intent.putExtra("post_id", newsList.get(position).getPost_id());
            intent.putExtra("post", newsList.get(position).getPost());
            intent.putExtra("title", newsList.get(position).getTitle());
            mContext.startActivity(intent);
        });

    }

    @Override
    public int getItemCount() {
        if (newsList != null)
            return newsList.size();
        else return 0;
    }

    class ViewHolder extends RecyclerView.ViewHolder {

        private TextView timeView, titleView, nameView;

        ViewHolder(View itemView) {
            super(itemView);

            timeView = itemView.findViewById(R.id.timeView);
            titleView = itemView.findViewById(R.id.titleView);
            nameView = itemView.findViewById(R.id.nameView);
        }
    }
}

File: app/src/main/java/com/konektedi/vs/news/NewsAdapter.kt
package com.konektedi.vs.news

import androidx.paging.PagedListAdapter
import android.content.Context
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import android.view.ViewGroup
import com.konektedi.vs.R
import com.konektedi.vs.utilities.common.NetworkState
import com.konektedi.vs.utilities.common.NetworkStateItemViewHolder
import com.konektedi.vs.utilities.models.Post

class NewsAdapter(private val mContext: Context, private val retryCallback: () -> Unit)
    : PagedListAdapter<Post, androidx.recyclerview.widget.RecyclerView.ViewHolder>(POST_COMPARATOR) {

    private var networkState: NetworkState? = null

    override fun onBindViewHolder(holder: androidx.recyclerview.widget.RecyclerView.ViewHolder, position: Int) {
        when (getItemViewType(position)) {
            R.layout.z_post -> (holder as NewsViewHolder).bind(getItem(position), mContext)
            R.layout.network_state_item -> (holder as NetworkStateItemViewHolder).bindTo(
                    networkState)
        }
    }

    override fun onBindViewHolder(
            holder: androidx.recyclerview.widget.RecyclerView.ViewHolder,
            position: Int,
            payloads: MutableList<Any>) {
        if (payloads.isNotEmpty()) {
            val item = getItem(position)
            (holder as NewsViewHolder).bind(item, mContext)
        } else {
            onBindViewHolder(holder, position)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): androidx.recyclerview.widget.RecyclerView.ViewHolder {
        return when (viewType) {
            R.layout.z_post -> NewsViewHolder.create(parent)
            R.layout.network_state_item -> NetworkStateItemViewHolder.create(parent, retryCallback)
            else -> throw IllegalArgumentException("unknown view type $viewType")
        }
    }

    private fun hasExtraRow() = networkState != null && networkState != NetworkState.LOADED

    override fun getItemViewType(position: Int): Int {
        return if (hasExtraRow() && position == itemCount - 1) {
            R.layout.network_state_item
        } else {
            R.layout.z_post
        }
    }

    override fun getItemCount(): Int {
        return super.getItemCount() + if (hasExtraRow()) 1 else 0
    }

    fun setNetworkState(newNetworkState: NetworkState?) {
        val previousState = this.networkState
        val hadExtraRow = hasExtraRow()
        this.networkState = newNetworkState
        val hasExtraRow = hasExtraRow()
        if (hadExtraRow != hasExtraRow) {
            if (hadExtraRow) {
                notifyItemRemoved(super.getItemCount())
            } else {
                notifyItemInserted(super.getItemCount())
            }
        } else if (hasExtraRow && previousState != newNetworkState) {
            notifyItemChanged(itemCount - 1)
        }
    }

    companion object {
        val POST_COMPARATOR = object : DiffUtil.ItemCallback<Post>() {
            override fun areContentsTheSame(oldItem: Post, newItem: Post): Boolean =
                    oldItem == newItem

            override fun areItemsTheSame(oldItem: Post, newItem: Post): Boolean =
                    oldItem.postId == newItem.postId
        }

    }
}
